examples/src/main/python/sql/streaming/structured_network_wordcount_session_window.py

```diff
@@ -0,0 +1,114 @@
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+r"""
+ split lines into words, group by words as key and use the state per key to track session of each key.
+
+ Usage: structured_network_wordcount_windowed.py <hostname> <port>
+ <hostname> and <port> describe the TCP server that Structured Streaming
+ would connect to receive data.
+
+ To run this on your local machine, you need to first run a Netcat server
+    `$ nc -lk 9999`
+ and then run the example
+    `$ bin/spark-submit
+    examples/src/main/python/sql/streaming/structured_network_wordcount_session_window.py
+    localhost 9999`
+"""
+import sys
+
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import explode
+from pyspark.sql.functions import split
+from pyspark.sql.functions import window
+from pyspark.sql.types import (
+    LongType,
+    StringType,
+    StructType,
+    StructField,
+    Row,
+)
+from pyspark.sql.streaming.state import GroupStateTimeout, GroupState
+import pandas as pd
+
+if __name__ == "__main__":
+    if len(sys.argv) != 3:
+        msg = ("Usage: structured_network_wordcount_session_window.py <hostname> <port>")
+        print(msg, file=sys.stderr)
+        sys.exit(-1)
+
+    host = sys.argv[1]
+    port = int(sys.argv[2])
+
+    spark = SparkSession\
+        .builder\
+        .appName("StructuredNetworkWordCountSessionWindow")\
+        .getOrCreate()
+
+    spark.sparkContext.setLogLevel("WARN")
+
+    # Create DataFrame representing the stream of input lines from connection to host:port
+    lines = spark\
+        .readStream\
+        .format('socket')\
+        .option('host', host)\
+        .option('port', port)\
+        .option('includeTimestamp', 'true')\
+        .load()
+
+    # Split the lines into words, retaining timestamps
+    # split() splits each line into an array, and explode() turns the array into multiple rows
+    events = lines.select(
+        explode(split(lines.value, ' ')).alias('sessionId'),
+        lines.timestamp.cast("long")
+    )
+
+    session_type = StructType(
+    [StructField("sessionId", StringType()), StructField("count", LongType()),
+    StructField("start", LongType()), StructField("end", LongType())]
+    )
+
+    def func(key, pdf_iter, state):
+        if state.hasTimedOut:
+            finished_session = state.get
+            state.remove()
+            yield pd.DataFrame({"sessionId": [finished_session[0]], "count": [finished_session[1]], "start": [finished_session[2]], "end": [finished_session[3]]})
+        else:
+            pdfs = list(pdf_iter)
```

Comment:
I know we won't encounter memory issue even we materialize all elements in iterator, but would be good to follow the traditional practice of using for-each. I don't think this will complicate the logic too much.

Reply:
Yes, this is an example so should better to have a best practice.

Reply:
Thanks, refactored.